File: python-algorithm/algo_08_graph/graph_2_dfs/L695-m.cpp
//
// Created by LiHeng on 2020/7/12.
//

#include <iostream>
#include <vector>
#include <cassert>
using namespace std;
class Solution{
private:
    int d[4][2] = {{0, 1},{1,0},{0,-1},{-1,0}} ;
    int m, n ;

    int dfs(vector<vector<int>>& grid, int x ,int y){
        if(grid[x][y] != 1)
            return 0;
        grid[x][y] = 0 ; //不会走重复路径
        int area = 1;
        for(int i = 0 ; i < 4 ; i++){
            int newx = x + d[i][0] ;
            int newy = y + d[i][1] ;
            if(newx>=0 && newx<m && newy>=0 && newy<n)
                area += dfs(grid, newx, newy);
        }
        return area;
    }

public:
    int maxAreaOfIsland(vector<vector<int>>& grid){
        m = grid.size() ;
        if(m == 0)
            return 0;
        n = grid[0].size() ;
        int res = 0 ;
        for(int i = 0 ; i < m ; ++i)
            for(int j = 0 ; j < n; ++j)
                if(grid[i][j] == 1){
                    res = max(res, dfs(grid , i , j));
                }
        return res ;
    }
};

//也可以下面这样写
class Solution1{
private:
    int d[4][2] = {{0, 1},{1,0},{0,-1},{-1,0}} ;
    int m , n ;
    vector<vector<bool>> vistited;


    int dfs(vector<vector<int>>& grid, int x ,int y){
        if(grid[x][y] != 1)
            return 0;
        vistited[x][y] = true ; //不会走重复路径
        int area = 1;
        for(int i = 0 ; i < 4 ; i++){
            int newx = x + d[i][0] ;
            int newy = y + d[i][1] ;
            if(newx>=0 && newx<m && newy>=0 && newy<n && !vistited[newx][newy])
                area += dfs(grid, newx, newy);
        }
        return area;
    }

public:
    int maxAreaOfIsland(vector<vector<int>>& grid){
        m = grid.size() ;
        if(m == 0)
            return 0;
        n = grid[0].size() ;
        int res = 0 ;
        vistited = vector<vector<bool>> (m, vector<bool>(n, false));
        for(int i = 0 ; i < m ; ++i)
            for(int j = 0 ; j < n; ++j)
                if(grid[i][j] == 1 && !vistited[i][j]){
                    res = max(res, dfs(grid , i , j));
                }
        return res ;
    }
};

File: python-algorithm/algo_08_graph/graph_2_dfs/L695-m.py
#--coding:utf-8--

# 有返回值的dfs
class Solution:
    def __init__(self, *args, **kwargs):
        self.m = 0
        self.n = 0
        self.oris = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    def InArea(self, x, y):
        return 0 <= x < self.m and 0 <= y < self.n

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.m = len(grid)
        self.n = len(grid[0])
        ans = 0
        for i in range(self.m):
            for j in range(self.n):
                if grid[i][j] == 1:
                    ans = max(ans, self.dfs(grid, i, j))
        return ans

    def dfs(self, grid, x, y):
        if grid[x][y] == 0: return 0
        res = 0
        # res = 1
        # 防止回头路
        grid[x][y] = 0
        for ori in self.oris:
            new_x = x + ori[0]
            new_y = y + ori[1]
            if self.InArea(new_x, new_y):
                res += self.dfs(grid, new_x, new_y)
        return res + 1
        # return res